File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/userCheckIn/UserCheckInHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.userCheckIn

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.userCheckIn.UserCheckInApi
import features.userCheckIn.models.CheckInResponseModel

import features.userCheckIn.serializeUserCheckInState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import features.userCheckIn.models.CheckInRequest

class UserCheckInHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/user_check_in"

    override val eventChannel: String
        get() = "customer_sdk/events/user_check_in"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return userCheckIn(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun userCheckIn(call: MethodCall): String? {
            val result = UserCheckInApi().userCheckIn(
                checkInRequest = Json.decodeFromString(call.argument<String>("checkInRequest") as String)
            ).toJsonResult<CheckInResponseModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        UserCheckInApi().userCheckInState.serializeUserCheckInState().observe { result ->
            pushSerializedResultToEventSink<UserCheckInApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/userCheckIn/UserCheckInHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class UserCheckInHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/user_check_in"

    var eventChannel: String = "customer_sdk/events/user_check_in"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await UserCheckInHandler.userCheckIn(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func userCheckIn(call: FlutterMethodCall) async -> String? {
        

        
        do {
        let result = try await UserCheckInApi().userCheckIn(methods: call.arguments as! [String : Any]).serializeUserCheckInApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
            }
        }
        catch {
            return nil
        }


        


        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        UserCheckInApi().userCheckInState
            .serializeUserCheckInState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
